[BUGFIX beta] allow watching of ES5+ Getter

stefanpenner opened the pull request:
Before this commit MandatorySetter enabled + current property being watched
resulted in `get` always fetching via meta.values. Although this typically works,
the default getter for a MandatorySetter already does this, so it is actually redundant.

Going further, when a user defines their own getter this code-path is incorrectly taken,
instead the original value (which originated from the getter) should be the consulted value.

This is blocking an ember-data fix, and likely causing grief for users who use ES5 getter/setters mixed in with our templates.

cc @krisselden

stefanpenner committed:
[BUGFIX Release] allow watching of ES5+ Getter

Before this commit MandatorySetter enabled + current property being watched
resulted in `get` always fetching via meta.values. Although this typically works,
the default getter for a MandatorySetter already does this, so it is actually redundant.

Going further, when a user defines their own getter this code-path is incorrectly taken,
instead the original value (which originated from the getter) should be the consulted value

stefanpenner committed:
[BUGFIX beta] unwatch should not destroy foreign descriptor

rwjblue commented:
Looks great! Can you add a test to https://github.com/emberjs/ember.js/blob/master/packages/ember-metal/tests/accessors/mandatory_setters_test.js also?

I think this should likely be bug fix beta to allow for our normal beta testing cycle.

stefanpenner commented:
> I think this should likely be bug fix beta to allow for our normal beta testing cycle.

This is breaking things in dev mode, and doesn't do anything radical to solve the issues. That being said, it may not be a critical fix on-our end.

@bmac is it alright if ember-data doesn't get this solution till 2.2?

rwjblue commented:
> is breaking things in dev mode, and doesn't do anything radical to solve the issues. That being said, it may not be a critical fix on-our end.

Yup, I understand that the bug exists in previously released versions (not just beta). Unless things are fundamentally broken in the wild (which would likely result in a large uproar) we should err on the side of using the normal beta process. This change is deep in the internals and a mistake in this area has the ability to cause massive issues that we may not be able to predict (we have had a number of [BUGFIX release] changes come back to bite us in the backside). Thankfully, this is **exactly** why we have a beta cycle.

stefanpenner commented:
@rwjblue yup just stating facts & checking with @bmac to see if this is critical or not to ed.

stefanpenner commented:
- [ ] fix more stuff, our tests here aren't very good..

stefanpenner committed:
[BUGFIX beta] Ember.set should not smash ES5 descriptor

[fixes 12480]

bmac commented:
@stefanpenner this functionality is not critical for Ember Data. We shipped a work around for this issue in 2.1 so we can refactor the workaround away in a future release when Ember's support improves.

stefanpenner committed:
[BUGFIX beta] remove redundant code (already the default M getter)

stefanpenner committed:
[BUGFIX beta] MandatorySetter: Restore properties correctly

* no longer rely on meta to store the value until next set after uninstalling MandatorySetter
* no longer incorrectly re-set on set to enumerable (enumerability should remain constant)
* test both enumerable / non-enumerable variants

stefanpenner committed:
[BUGFIX beta] MandatorySetter should work with inheritance.

stefanpenner commented:
@bmac I think this addresses everything I could think of, if you have time to confirm on your end that would be good.

@rwjblue r?
@krisselden r?

stefanpenner commented:
- [ ] https://github.com/emberjs/ember.js/commit/b66a13f56b5607aa91dca4862da9028078eab7a0 removed keys_test and moved them, I added some to that file, i should move them aswell

mixonic commented:
So many new tests :clap:

stefanpenner commented:
Test failures seem related to phantom, will investigate soon

rwjblue commented on packages/ember-metal/lib/properties.js:
```diff
@@ -38,6 +41,16 @@ export function DEFAULT_GETTER_FUNCTION(name) {
 
+export function INHERITING_GETTER_FUNCTION(name) {
+  function IGETTER_FUNCTION() {
+    var proto = Object.getPrototypeOf(this);
```
It seems to me (just off the bat) that we need to do the same thing as was added in https://github.com/emberjs/ember.js/pull/12314 to lookup the getter through the entire prototype chain.  As it exists here, I believe that a getter from a parent class will not be found.

Specifically, I think we should use [lookupDescriptor](https://github.com/emberjs/ember.js/blob/bb5cc203c79da47af1ead291848c280c63ee6fe5/packages/ember-metal/lib/watch_key.js#L40-L53), and add some tests around this behavior like [these ones](https://github.com/emberjs/ember.js/blob/bb5cc203c79da47af1ead291848c280c63ee6fe5/packages/ember-metal/tests/accessors/mandatory_setters_test.js#L97-L153) only for getter.

stefanpenner replied:
I don't think so, this exists to simulate the behavior if no descriptor existed at this level. It aims to re-target get one step up the prototype chain.

Related tests: https://github.com/emberjs/ember.js/pull/12491/files#diff-25946da80206a01a2f680ad806bb1780R370

stefanpenner replied:
wait i believe i see it now.

stefanpenner replied:
adding another test

stefanpenner replied:
Added a test, will investigate another time (time for bed)

stefanpenner commented:
After chatting with kris, he had a good idea, installing the mandatory setter for inherited properties in at the level defined (instead of at the leaf). Will explore this.

stefanpenner commented:
@rwjblue @krisselden r?

stefanpenner commented:
I don't believe the IE9 failure is critical, but I will investigate it once I have IE9 handy. It appears to just be a IE9ism, and very specific tests one my part.

stefanpenner commented:
IE9 issue is fixed in my VM, lets see what sauce thinks.

stefanpenner commented:
![](http://i.kinja-img.com/gawker-media/image/upload/s--2iox6xgo--/bgkyeggoyscnn41ryita.jpg)

stefanpenner commented on packages/ember-metal/lib/property_set.js:
```diff
@@ -69,23 +73,25 @@ export function set(obj, keyName, value, tolerant) {
           } else {
-            meta.writeValues(keyName, value);
+            let descriptor = lookupDescriptor(obj, keyName);
+            let isMandatorySetter = descriptor && descriptor.set && descriptor.set.isMandatorySetter;
```
We should strip this out if `isMandatorySetter` is disabled.

stefanpenner commented:
It is worth noting, this will make development builds of ember run slower (but will leave prod builds the same). Unfortunately short of either taking a drastically different approach, or disabling mandatory setter entirely, I'm unsure of another approach.

It is possible for us to explore micro-related improvements, but...
